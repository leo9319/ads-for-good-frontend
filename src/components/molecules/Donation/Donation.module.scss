@import '../../../styles/colours';
@import '../../../styles/typography';

.donation-component {
  background-color: $color-white;
  border-radius: 16px;
  padding: 16px;
  width: 100%;
  max-width: 328px;
  box-shadow: 0px 4px 24px $color-box-shadow;

  display: flex;
  flex-direction: column;
  align-items: center;
  gap: 24px;
  box-sizing: border-box;

  .content-container {
    display: flex;
    flex-direction: column;
    align-items: center;
    width: 100%;
    box-sizing: border-box;
    gap: 24px;
  }

  .title-container {
    align-self: center;
    text-align: center;
    width: 100%;
  }

  .title {
    font-family: var(--typography-default-font-family);
    font-weight: var(--typography-font-weight-semibold);
    color: $color-neutral-950;
    width: 100%;
  }

  .form-container {
    display: flex;
    flex-direction: column;
    gap: 16px;
    margin: auto;
    width: 100%;
  }

  .frequency-group,
  .amount-group {
    display: flex;
    gap: 8px;
    width: 100%;
  }

  .toggle-item {
    flex: 1;
    min-width: 80px;
    width: var(--toggle-width);
    height: var(--toggle-height);
    padding: var(--toggle-padding);

    font-family: var(--typography-default-font-family);
    font-weight: var(--typography-font-weight-regular);
    font-size: var(--typography-font-size-4);
    line-height: var(--typography-line-height-4);
    text-align: center;

    border-radius: 12px;
    border: 0px solid $color-neutral-200;
    background-color: var(--Background-secondary, #eaebee);
    color: var(--Text-primary, #17191f);
    cursor: pointer;
    transition:
      background-color 0.2s ease,
      color 0.2s ease;

    &[data-state='on'] {
      background-color: var(--Background-primary, #17191f);
      color: var(--Text-primary-inverse, #ffffff);
    }
  }

  .custom-input {
    width: 100%;
    margin-bottom: 0;
    font-size: var(--typography-font-size-3);
    color: $color-neutral-900;

    &::placeholder {
      color: $color-neutral-500;
    }
  }

  .button-container {
    display: flex;
    align-items: center;
    justify-content: center;
    width: 100%;
  }

  .donate-button {
    width: 100%;
    border-radius: 60px;
    background-color: var(--Button-Primary-default, #1049a9);
    color: $color-white;

    display: flex;
    justify-content: center;
    align-items: center;
    overflow: hidden;

    svg {
      width: 20px;
      height: 20px;
      flex-shrink: 0;
    }

    span {
      font-family: var(--font-family-heading);
      font-weight: 400;
      letter-spacing: 0%;
      color: var(--Text-primary-inverse, #ffffff);
      white-space: nowrap;
    }

    &.sm-md {
      span {
        font-size: var(--font-size-heading-xs);
        line-height: var(--line-height-heading-xs);
      }
    }

    &.lg-xl {
      span {
        font-size: var(--font-size-heading-sm);
        line-height: var(--line-height-heading-sm);
      }
    }
  }

  &.sm-md {
    --toggle-width: 93.3333px;
    --toggle-height: 44px;
    --toggle-padding: 10px 16px;

    max-width: 328px;
    padding: 16px;
    border-radius: 16px;

    .content-container {
      width: 100%;
      gap: 24px;
    }

    .title-container,
    .form-container,
    .button-container {
      width: 100%;
    }

    .title-container {
      margin-bottom: 8px;

      .title {
        font-size: var(--typography-font-size-6);
        line-height: var(--typography-line-height-6);
      }
    }

    .form-container {
      gap: 16px;
    }
  }

  &.lg-xl {
    --toggle-width: 111.3333px;
    --toggle-height: 44px;
    --toggle-padding: 10px 16px;

    max-width: 430px;
    padding: 40px;
    border-radius: 32px;

    .content-container {
      width: 100%;
      gap: 24px;
    }

    .title-container,
    .form-container,
    .button-container {
      width: 100%;
    }

    .title-container {
      margin-bottom: 8px;

      .title {
        max-width: 100%;
        font-size: var(--typography-font-size-7);
        line-height: var(--typography-line-height-7);
      }
    }

    .form-container {
      gap: 16px;
    }
  }
}
