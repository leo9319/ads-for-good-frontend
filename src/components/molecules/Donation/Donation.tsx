import React, { useState } from 'react';
import styles from './Donation.module.scss';
import Input from '../../atoms/Input';
import Button from '../../../radix/atoms/Button';
import Text, { TextProps as CustomTextProps } from '../../../radix/atoms/Text';
import * as ToggleGroup from '@radix-ui/react-toggle-group';
import { classNames } from '@utils/common/classNames';

export interface DonationProps {
  /**
   * Title configuration for the donation block.
   * Defaults to: { text: 'Choose your donation' }
   */
  titleConfig?: Partial<CustomTextProps> & { text: string };

  /**
   * Donation frequency options (e.g., ["Monthly", "One-time"]).
   * Required.
   */
  frequencies: string[];

  /**
   * Preset donation amount options (e.g., [10, 50, 100]).
   * Required.
   */
  amounts: number[];

  /**
   * Optional default value for the custom amount input.
   * Defaults to: 0
   */
  defaultCustomAmount?: number;

  /**
   * Callback invoked when the donate button is clicked.
   * Returns selected amount and frequency.
   */
  onDonate: (amount: number, frequency: string) => void;

  /**
   * Props to configure the Donate button (text, size, effects, etc.)
   */
  buttonProps?: {
    text: string;
    size?: 'sm' | 'md' | 'lg' | 'xl';
    mode?: 'primary' | 'secondary';
    effects?: 'none' | 'special';
    icon?: string;
    iconLeft?: boolean;
  };

  /**
   * Size of the entire component layout.
   * "sm-md" (default) for compact layouts, "lg-xl" for spacious ones.
   */
  size?: 'sm-md' | 'lg-xl';
}

export const Donation: React.FC<DonationProps> = ({
  titleConfig = { text: 'Choose your donation' },
  frequencies,
  amounts,
  defaultCustomAmount = 0,
  onDonate,
  buttonProps,
  size = 'sm-md',
}) => {
  // Set default selected frequency and amount from the first options
  const [selectedFrequency, setSelectedFrequency] = useState(frequencies[0]);
  const [selectedAmount, setSelectedAmount] = useState<number | null>(
    amounts[0]
  );

  // Populate custom amount field if defaultCustomAmount > 0
  const [customAmount, setCustomAmount] = useState<string>(
    defaultCustomAmount > 0 ? defaultCustomAmount.toString() : ''
  );

  const handleDonate = () => {
    const amount = selectedAmount ?? (parseFloat(customAmount) || 0);
    if (amount > 0) onDonate(amount, selectedFrequency);
  };

  const handleCustomAmountChange = (e: React.ChangeEvent<HTMLInputElement>) => {
    const value = e.target.value;
    setCustomAmount(value);

    const numericValue = parseFloat(value);
    // If input matches a preset amount, highlight the corresponding toggle
    if (!isNaN(numericValue) && amounts.includes(numericValue)) {
      setSelectedAmount(numericValue);
    } else {
      setSelectedAmount(null); // custom value
    }
  };

  return (
    <div className={classNames(styles['donation-component'], styles[size])}>
      <div className={styles['content-container']}>
        {/* Title */}
        <div className={styles['title-container']}>
          <Text
            as="p"
            {...titleConfig}
            className={classNames(styles.title, titleConfig?.className)}
          >
            {titleConfig.text}
          </Text>
        </div>

        {/* Frequency and amount selection */}
        <div className={styles['form-container']}>
          {/* Frequency toggle */}
          <ToggleGroup.Root
            type="single"
            value={selectedFrequency}
            onValueChange={val => val && setSelectedFrequency(val)}
            className={styles['frequency-group']}
          >
            {frequencies.map(freq => (
              <ToggleGroup.Item
                key={freq}
                value={freq}
                className={styles['toggle-item']}
              >
                {freq}
              </ToggleGroup.Item>
            ))}
          </ToggleGroup.Root>

          {/* Amount toggle */}
          <ToggleGroup.Root
            type="single"
            value={selectedAmount?.toString() || ''}
            onValueChange={val => {
              if (val) {
                const numericVal = Number(val);
                setSelectedAmount(numericVal);
                setCustomAmount(val);
              }
            }}
            className={styles['amount-group']}
          >
            {amounts.map(amt => (
              <ToggleGroup.Item
                key={amt}
                value={amt.toString()}
                className={styles['toggle-item']}
              >
                ${amt}
              </ToggleGroup.Item>
            ))}
          </ToggleGroup.Root>

          {/* Custom amount input */}
          <Input
            name="customAmount"
            type="number"
            min="0"
            step="1"
            placeholder="Custom amount"
            value={customAmount === '' ? undefined : customAmount}
            onChange={handleCustomAmountChange}
            className={styles['custom-input']}
          />
        </div>

        {/* CTA Button */}
        <div className={styles['button-container']}>
          <Button
            className={classNames(styles['donate-button'], styles[size])}
            onClick={handleDonate}
            {...{
              size: 'none',
              mode: 'primary',
              effects: 'special',
              ...buttonProps,
            }}
          >
            <span>{buttonProps?.text || 'Donate Now'}</span>
          </Button>
        </div>
      </div>
    </div>
  );
};

export default Donation;
