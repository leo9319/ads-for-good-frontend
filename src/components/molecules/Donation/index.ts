export { Donation } from './Donation';
export type { DonationProps } from './Donation';
