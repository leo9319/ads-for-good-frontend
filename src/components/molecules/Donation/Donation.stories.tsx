import { Meta, StoryObj } from '@storybook/react';
import { action } from '@storybook/addon-actions';
import { Donation, DonationProps } from './Donation';

const meta: Meta<typeof Donation> = {
  title: 'Library/Molecules/Donation',
  component: Donation,
  tags: ['autodocs'],
  parameters: {
    layout: 'centered',
  },
  argTypes: {
    titleConfig: {
      control: { type: 'object' },
      defaultValue: { text: 'Support Our Mission', size: '5', weight: 'bold' },
    },
    frequencies: {
      control: { type: 'object' },
      defaultValue: ['One-time', 'Monthly'],
    },
    amounts: {
      control: { type: 'object' },
      defaultValue: [500, 1000, 1500],
    },
    defaultCustomAmount: {
      control: { type: 'number' },
      defaultValue: 0,
    },
    onDonate: {
      action: 'donate-submitted',
    },
    buttonProps: {
      control: { type: 'object' },
      defaultValue: {
        text: 'Donate Now',
        size: 'xl',
        mode: 'primary',
        effects: 'special',
        icon: 'icon-arrow',
        iconLeft: false,
      },
    },
    size: {
      control: { type: 'radio' },
      options: ['sm-md', 'lg-xl'],
      defaultValue: 'sm-md',
    },
  },
};

export default meta;

type DonationStory = StoryObj<typeof Donation>;

const defaultArgs: DonationProps = {
  titleConfig: {
    text: 'Choose your donation',
    size: '5',
    weight: 'bold',
  },
  frequencies: ['One-time', 'Monthly'],
  amounts: [25, 50, 100],
  defaultCustomAmount: 0,
  onDonate: action('donate-submitted'),
  buttonProps: {
    text: 'Donate Now',
    size: 'xl',
    mode: 'primary',
    effects: 'special',
    icon: 'icon-arrow',
    iconLeft: false,
  },
  size: 'sm-md',
};

export const Default: DonationStory = {
  args: {
    ...defaultArgs,
  },
};
