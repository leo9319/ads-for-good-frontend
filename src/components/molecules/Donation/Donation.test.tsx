import React from 'react';
import { render, screen, fireEvent } from '@testing-library/react';
import '@testing-library/jest-dom';
import { Donation, DonationProps } from './Donation';

describe('Donation Component', () => {
  const mockOnDonate = jest.fn();

  const defaultProps: DonationProps = {
    titleConfig: {
      text: 'Support Our Mission',
    },
    frequencies: ['One-time', 'Monthly'],
    amounts: [500, 1000, 1500],
    onDonate: mockOnDonate,
    buttonProps: {
      text: 'Donate Now',
    },
  };

  beforeEach(() => {
    mockOnDonate.mockClear();
  });

  it('renders title and options correctly', () => {
    render(<Donation {...defaultProps} />);
    expect(screen.getByText('Support Our Mission')).toBeInTheDocument();
    expect(screen.getByText('One-time')).toBeInTheDocument();
    expect(screen.getByText('Monthly')).toBeInTheDocument();
    expect(screen.getByText('$500')).toBeInTheDocument();
    expect(screen.getByPlaceholderText('Custom amount')).toBeInTheDocument();
  });

  it('calls onDonate with selected amount and frequency', () => {
    render(<Donation {...defaultProps} />);
    fireEvent.click(screen.getByText('$1000'));
    fireEvent.click(screen.getByText('Donate Now'));
    expect(mockOnDonate).toHaveBeenCalledWith(1000, 'One-time');
  });

  it('calls onDonate with custom amount if entered', () => {
    render(<Donation {...defaultProps} />);
    fireEvent.change(screen.getByPlaceholderText('Custom amount'), {
      target: { value: '777' },
    });
    fireEvent.click(screen.getByText('Donate Now'));
    expect(mockOnDonate).toHaveBeenCalledWith(777, 'One-time');
  });

  it('updates input field when toggle amount is selected', () => {
    render(<Donation {...defaultProps} />);
    fireEvent.click(screen.getByText('$1500'));
    const input = screen.getByPlaceholderText(
      'Custom amount'
    ) as HTMLInputElement;
    expect(input.value).toBe('1500');
  });

  it('clears toggle selection when custom amount does not match', () => {
    render(<Donation {...defaultProps} />);
    fireEvent.click(screen.getByText('$1000'));
    const input = screen.getByPlaceholderText('Custom amount');
    fireEvent.change(input, { target: { value: '888' } });
    fireEvent.click(screen.getByText('Donate Now'));
    expect(mockOnDonate).toHaveBeenCalledWith(888, 'One-time');
  });

  it('switches frequency correctly', () => {
    render(<Donation {...defaultProps} />);
    fireEvent.click(screen.getByText('Monthly'));
    fireEvent.click(screen.getByText('Donate Now'));
    expect(mockOnDonate).toHaveBeenCalledWith(500, 'Monthly');
  });

  it('shows placeholder in custom input when empty', () => {
    render(<Donation {...defaultProps} />);
    const input = screen.getByPlaceholderText(
      'Custom amount'
    ) as HTMLInputElement;
    expect(input).toBeInTheDocument();
    expect(input.value).toBe('');
  });
});

describe('Donation Component Snapshot', () => {
  const mockOnDonate = jest.fn();

  const snapshotProps: DonationProps = {
    titleConfig: {
      text: 'Support Our Mission',
    },
    frequencies: ['One-time', 'Monthly'],
    amounts: [500, 1000, 1500],
    onDonate: mockOnDonate,
    buttonProps: {
      text: 'Donate Now',
    },
  };

  it('matches snapshot', () => {
    const { asFragment } = render(<Donation {...snapshotProps} />);
    expect(asFragment()).toMatchSnapshot();
  });
});
