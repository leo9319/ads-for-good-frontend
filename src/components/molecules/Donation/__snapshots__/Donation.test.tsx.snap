// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Donation Component Snapshot matches snapshot 1`] = `
<DocumentFragment>
  <div
    class="donation-component sm-md"
  >
    <div
      class="content-container"
    >
      <div
        class="title-container"
      >
        <p
          class="rt-Text mocked-class-name typography-font-family-default title"
          id=""
          text="Support Our Mission"
        >
          Support Our Mission
        </p>
      </div>
      <div
        class="form-container"
      >
        <div
          class="frequency-group"
          dir="ltr"
          role="group"
          style="outline: none;"
          tabindex="0"
        >
          <button
            aria-checked="true"
            class="toggle-item"
            data-radix-collection-item=""
            data-state="on"
            role="radio"
            tabindex="-1"
            type="button"
          >
            One-time
          </button>
          <button
            aria-checked="false"
            class="toggle-item"
            data-radix-collection-item=""
            data-state="off"
            role="radio"
            tabindex="-1"
            type="button"
          >
            Monthly
          </button>
        </div>
        <div
          class="amount-group"
          dir="ltr"
          role="group"
          style="outline: none;"
          tabindex="0"
        >
          <button
            aria-checked="true"
            class="toggle-item"
            data-radix-collection-item=""
            data-state="on"
            role="radio"
            tabindex="-1"
            type="button"
          >
            $500
          </button>
          <button
            aria-checked="false"
            class="toggle-item"
            data-radix-collection-item=""
            data-state="off"
            role="radio"
            tabindex="-1"
            type="button"
          >
            $1000
          </button>
          <button
            aria-checked="false"
            class="toggle-item"
            data-radix-collection-item=""
            data-state="off"
            role="radio"
            tabindex="-1"
            type="button"
          >
            $1500
          </button>
        </div>
        <div
          class="formInput custom-input"
        >
          <input
            id="customAmount"
            min="0"
            placeholder="Custom amount"
            step="1"
            type="number"
            value=""
          />
        </div>
      </div>
      <div
        class="button-container"
      >
        <button
          aria-label="Donate Now"
          class="rt-reset rt-BaseButton rt-r-size-2 rt-variant-solid rt-Button wvButton xl primary special donate-button sm-md"
          data-accent-color=""
        >
          <div
            class="buttonContent textOnly"
          >
            <span>
              Donate Now
            </span>
          </div>
        </button>
      </div>
    </div>
  </div>
</DocumentFragment>
`;
